core/trino-main/src/main/java/io/trino/execution/scheduler/EventDrivenFaultTolerantQueryScheduler.java

```diff
@@ -2174,7 +2221,82 @@ private Optional<OutputDataSizeEstimateResult> getEstimatedOutputDataSize()
             return Optional.of(new OutputDataSizeEstimateResult(new OutputDataSizeEstimate(estimateBuilder.build()), OutputDataSizeEstimateStatus.ESTIMATED_BY_PROGRESS));
         }
 
-        private Optional<OutputDataSizeEstimateResult> getEstimatedSmallStageOutputDataSize(Function<StageId, StageExecution> stageExecutionLookup)
+        private Optional<OutputDataSizeEstimateResult> getEstimatedMarkDistinctStageOutputDataSize(Function<StageId, StageExecution> stageExecutionLookup, int depth)
+        {
+            if (!markDistinctChainEstimationEnabled) {
+                return Optional.empty();
+            }
+
+            if (depth >= markDistinctChainEstimationMaxDepth) {
+                return Optional.empty();
+            }
+
+            PlanFragment planFragment = this.getStageInfo().getPlan();
+            if (!isSingleMarkDistinctStage(planFragment)) {
+                return Optional.empty();
+            }
+
+            List<RemoteSourceNode> remoteSourceNodes = planFragment.getRemoteSourceNodes();
+            long inputSizeEstimate = 0;
+            for (RemoteSourceNode remoteSourceNode : remoteSourceNodes) {
+                for (PlanFragmentId sourceFragmentId : remoteSourceNode.getSourceFragmentIds()) {
+                    StageId sourceStageId = StageId.create(queryStateMachine.getQueryId(), sourceFragmentId);
+
+                    StageExecution sourceStage = stageExecutionLookup.apply(sourceStageId);
+                    Optional<OutputDataSizeEstimateResult> sourceStageOutputDataSize = sourceStage.getOutputDataSize(stageExecutionLookup, false, depth + 1);
+
+                    if (sourceStageOutputDataSize.isEmpty()) {
+                        // We need estimate from all sources
+                        return Optional.empty();
+                    }
+
+                    inputSizeEstimate += sourceStageOutputDataSize.orElseThrow().getOutputDataSizeEstimate().getTotalSizeInBytes();
+                }
+            }
+
+            inputSizeEstimate = (long) (inputSizeEstimate * 1.1); // conservatively assume some data growth
```

Comment:
You will never output less data. As you need to output all the columns you got + marker column, right?

Reply:
see the example below.
`hashvalue_10` is dropped by projection on top of mark distinct because its not needed anymore. The stage produces `115.33GB` on `134.11GB` input
```
 Fragment 3 [HASH]                                                                                                                                                                                                                                   
     CPU: 31.31m, Scheduled: 35.11m, Blocked 2.49h (Input: 1.36h, Output: 14.14m), Input: 2879987999 rows (134.11GB); per task: avg.: 479997999.83 std.dev.: 24562769.56, Output: 2879987999 rows (115.33GB)                                         
     Output buffer active time: 3.51m, buffer utilization distribution (%): {p01=0.00, p05=0.00, p10=0.28, p25=13.20, p50=48.25, p75=86.20, p90=100.93, p95=104.07, p99=111.14, max=332.50}                                                          
     Task output distribution: {count=6.00, p01=18.75GB, p05=18.75GB, p10=18.75GB, p25=18.75GB, p50=18.80GB, p75=18.85GB, p90=21.42GB, p95=21.42GB, p99=21.42GB, max=21.42GB}                                                                        
     Task input distribution: {count=6.00, p01=21.80GB, p05=21.80GB, p10=21.80GB, p25=21.80GB, p50=21.86GB, p75=21.92GB, p90=24.91GB, p95=24.91GB, p99=24.91GB, max=24.91GB}                                                                         
     Output layout: [ss_customer_sk, ss_promo_sk, ss_quantity, ss_customer_sk$distinct, $hashvalue_8, $hashvalue_9]                                                                                                                                  
     Output partitioning: HASH [ss_quantity, ss_promo_sk][$hashvalue_9]                                                                                                                                                                              
     Project[]                                                                                                                                                                                                                                       
     │   Layout: [ss_customer_sk:bigint, ss_promo_sk:bigint, ss_quantity:integer, $hashvalue_8:bigint, $hashvalue_9:bigint, ss_customer_sk$distinct:boolean]                                                                                         
     │   Estimates: {rows: ? (?), cpu: ?, memory: 0B, network: 0B}                                                                                                                                                                                   
     │   CPU: 4.65m (9.23%), Scheduled: 5.29m (8.12%), Blocked: 14.14m (2.73%), Output: 2879987999 rows (115.33GB)                                                                                                                                   
     │   metrics:                                                                                                                                                                                                                                    
     │     'CPU time distribution (s)' = {count=192, p01=0.45, p05=0.46, p10=0.47, p25=0.48, p50=0.52, p75=0.54, p90=0.59, p95=0.64, p99=0.65, min=0.45, max=0.65}                                                                                   
     │     'Input rows distribution' = {count=192, p01=14311071.00, p05=14314600.00, p10=14318298.20, p25=14324303.83, p50=14334351.31, p75=14972647.68, p90=14985128.30, p95=15619623.00, p99=16268002.00, min=14308571.00, max=79751713.00}        
     │     'Projection CPU time' = {duration=6.04s}                                                                                                                                                                                                  
     │     'Scheduled time distribution (s)' = {count=192, p01=0.51, p05=0.53, p10=0.54, p25=0.58, p50=0.62, p75=0.67, p90=0.78, p95=0.81, p99=0.89, min=0.51, max=0.90}                                                                             
     │   Input avg.: 14999937.49 rows, Input std.dev.: 31.36%                                                                                                                                                                                        
     └─ MarkDistinct[distinct = [ss_quantity:integer, ss_customer_sk:bigint], marker = ss_customer_sk$distinct, hash = [$hashvalue_10]]                                                                                                              
        │   Layout: [ss_customer_sk:bigint, ss_promo_sk:bigint, ss_quantity:integer, $hashvalue_8:bigint, $hashvalue_9:bigint, $hashvalue_10:bigint, ss_customer_sk$distinct:boolean]                                                                
        │   CPU: 22.51m (44.75%), Scheduled: 24.49m (37.55%), Blocked: 0.00ns (0.00%), Output: 2879987999 rows (139.47GB)                                                                                                                            
        │   metrics:                                                                                                                                                                                                                                 
        │     'CPU time distribution (s)' = {count=192, p01=6.23, p05=6.29, p10=6.34, p25=6.85, p50=7.11, p75=7.26, p90=7.42, p95=7.45, p99=7.57, min=6.23, max=11.58}                                                                               
        │     'Input rows distribution' = {count=192, p01=14311071.00, p05=14314600.00, p10=14318298.20, p25=14324303.83, p50=14334351.31, p75=14972647.68, p90=14985128.30, p95=15619623.00, p99=16268002.00, min=14308571.00, max=79751713.00}     
        │     'Scheduled time distribution (s)' = {count=192, p01=6.58, p05=6.70, p10=6.77, p25=7.32, p50=7.80, p75=7.99, p90=8.13, p95=8.21, p99=8.35, min=6.58, max=12.64}                                                                         
        │   Input avg.: 14999937.49 rows, Input std.dev.: 31.36%                                                                                                                                                                                     
        └─ LocalExchange[partitioning = HASH, hashColumn = [$hashvalue_10], arguments = ["ss_quantity", "ss_customer_sk"]]                                                                                                                           
           │   Layout: [ss_customer_sk:bigint, ss_promo_sk:bigint, ss_quantity:integer, $hashvalue_8:bigint, $hashvalue_9:bigint, $hashvalue_10:bigint]                                                                                              
           │   Estimates: {rows: 3026764054 (127.46GB), cpu: 127.46G, memory: 0B, network: 0B}                                                                                                                                                       
           │   CPU: 3.33m (6.62%), Scheduled: 4.04m (6.20%), Blocked: 55.06m (10.62%), Output: 2879987999 rows (134.11GB)                                                                                                                            
           │   metrics:                                                                                                                                                                                                                              
           │     'Blocked time distribution (s)' = {count=192, p01=1.63, p05=1.68, p10=1.71, p25=2.37, p50=3.17, p75=3.96, p90=9.63, p95=9.67, p99=9.72, min=1.63, max=9.73}                                                                         
           │     'CPU time distribution (s)' = {count=192, p01=0.44, p05=0.54, p10=0.58, p25=0.64, p50=0.76, p75=0.97, p90=1.16, p95=1.30, p99=1.61, min=0.43, max=1.68}                                                                             
           │     'Input rows distribution' = {count=192, p01=8109041.00, p05=9810245.00, p10=10467719.00, p25=11635394.43, p50=13708553.78, p75=17788937.45, p90=21604626.60, p95=24114791.00, p99=30104839.00, min=7974017.00, max=31100291.00}     
           │     'Scheduled time distribution (s)' = {count=192, p01=0.47, p05=0.58, p10=0.63, p25=0.71, p50=0.83, p75=1.07, p90=1.30, p95=1.51, p99=1.80, min=0.44, max=1.88}                                                                       
           │   Input avg.: 14999937.49 rows, Input std.dev.: 29.62%                                                                                                                                                                                  
           └─ RemoteSource[sourceFragmentIds = [4]]                                                                                                                                                                                                  
                  Layout: [ss_customer_sk:bigint, ss_promo_sk:bigint, ss_quantity:integer, $hashvalue_11:bigint, $hashvalue_12:bigint, $hashvalue_13:bigint]                                                                                         
                  CPU: 38.18s (1.26%), Scheduled: 53.45s (1.37%), Blocked: 1.36h (15.75%), Output: 2879987999 rows (134.11GB)                                                                                                                        
                  metrics:                                                                                                                                                                                                                           
                    'Blocked time distribution (s)' = {count=192, p01=18.05, p05=18.18, p10=18.34, p25=25.20, p50=26.37, p75=28.23, p90=28.64, p95=28.75, p99=28.83, min=18.00, max=28.91}                                                           
                    'CPU time distribution (s)' = {count=192, p01=0.12, p05=0.13, p10=0.13, p25=0.15, p50=0.18, p75=0.23, p90=0.29, p95=0.33, p99=0.50, min=0.12, max=0.52}                                                                          
                    'Input rows distribution' = {count=192, p01=8109041.00, p05=9810245.00, p10=10467719.00, p25=11635394.43, p50=13708553.78, p75=17788937.45, p90=21604626.60, p95=24114791.00, p99=30104839.00, min=7974017.00, max=31100291.00}  
                    'Scheduled time distribution (s)' = {count=192, p01=0.15, p05=0.17, p10=0.18, p25=0.20, p50=0.25, p75=0.32, p90=0.41, p95=0.53, p99=0.68, min=0.15, max=0.71}                                                                    
                  Input avg.: 14999937.49 rows, Input std.dev.: 29.62%   
```

Reply:
You are right - I will not implement it though. It is very rough estimatin anyway - it will not buy us much if we just look at that. In the same spirit we should at least look at number of columns at types - but it would complicate logic a lot.